godot-macros/src/class/godot_api.rs

```diff
@@ -229,11 +232,29 @@ fn process_godot_fns(decl: &mut Impl) -> Result<(Vec<FuncDefinition>, Vec<Functi
                 return attr.bail("generic fn parameters are not supported", method);
             }
 
-            match attr.ty {
-                BoundAttrType::Func { rename } => {
+            match &attr.ty {
+                BoundAttrType::Func {
+                    rename,
+                    has_gd_self,
+                } => {
                     // Signatures are the same thing without body
-                    let sig = util::reduce_to_signature(method);
-                    func_definitions.push(FuncDefinition { func: sig, rename });
+                    let mut sig = util::reduce_to_signature(method);
+                    if *has_gd_self {
+                        if sig.params.is_empty() {
+                            return attr.bail("When the attribute gd_self is used the method must have a first argument whose type is Gd<Self>", method);
+                        }
+                        sig.params[0].0 = FnParam::Receiver(FnReceiverParam {
+                            attributes: vec![],
+                            tk_ref: Some(proc_macro2::Punct::new('&', proc_macro2::Spacing::Alone)),
+                            tk_mut: None,
+                            tk_self: Ident::new("self", proc_macro2::Span::call_site()),
+                        });
```

Comment:
You construct a `&self` parameter here, but what is this used for?
And how do you know it's not `&mut self`?

Reply:
This is an attempt to simplify `func.rs`'s `get_signature_info` method. Now there's no case where both `gd_self` being true and the first argument being `&self` or `&mut self` can conflict set `receiver_type` twice.
I don't love this, and could also go back and just add a panic! to `get_signature_info` if the first argument is `&self`.

Reply:
Thinking on it again, there's already a panic! in `get_signature_info`.
I think it makes sense to inelegantly handle the case where `gd_self` is true and the first argument is `&self` in `get_signature_info` with a panic! as a fallback. And then have `process_godot_fns` check for that case and emit a prettier looking bail error instead.

I'll upload a separate commit that switches to that, so it'll be easier to go with which you prefer.